src/AppBundle/Model/Services/UserService.php

```diff
@@ -0,0 +1,169 @@
+use Ucsf\LdapOrmBundle\Repository\Repository;
+
+class UserService implements UserProviderInterface
+{
+
+    /**
+     * A Referec to the LdapEntityService to work with the ldap
+     *
+     * @var LdapEntityManager
+     */
+    private $ldapEntityManager;
+
+    /**
+     * Just the logger
+     *
+     * @var Logger
+     */
+    private $logger;
+
+    /**
+     * The ldapManager of the LDAPBundle
+     *
+     * @var Logger
+     * @var LocaleDataGenerator
+     */
+    public function __construct(Logger $logger, LdapEntityManager $ldapEntityManager)
+    {
+        $this->ldapEntityManager = $ldapEntityManager;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Searches the PbnlAccount (ldap) and returns a User
+     * Find by GivenName
+     *
+     * @param string $givenName
+     * @return User
+     */
+    public function getUserByGivenName(String $givenName)
+    {
+        /** @var Repository $pbnlAccountRepository */
+        $pbnlAccountRepository = $this->ldapEntityManager->getRepository(PbnlAccount::class);
+
+        /** @var PbnlAccount $ldapPbnlAccount */
+        $ldapPbnlAccount = $pbnlAccountRepository->findOneByGivenName($givenName);
+
+        if (count($ldapPbnlAccount) == 0) {
+            throw new UsernameNotFoundException(
+                sprintf('Username "%s" does not exist.', $givenName)
+            );
+        }
+
+        $user = $this->entitiesToUser($ldapPbnlAccount);
+
+        return $user;
+    }
+
+    /**
+     * Creates a User with the data of a PbnlAccount (ldap)
+     * and later mayby a MySqlPbnlAccount
+     *
+     * @param PbnlAccount $ldapPbnlAccount
+     * @return User
+     */
+    private function entitiesToUser(PbnlAccount $ldapPbnlAccount)
+    {
+        // skip the "{SSHA}"
+        $b64 = substr($ldapPbnlAccount->getUserPassword(), 6);
+
+        // base64 decoded
+        $b64_dec = base64_decode($b64);
+
+        // the salt (given it is a 8byte one)
+        $salt = substr($b64_dec, -8);
+        // the sha1 part
+        $hashedPassword = substr($b64_dec, 0, 20);
+
+        //Fill up the user
+        $user = new User($ldapPbnlAccount->getGivenName(), $hashedPassword, $salt, []);
+        $user->setDn($ldapPbnlAccount->getDn());
+        $user->setCity($ldapPbnlAccount->getL());
+        $user->setFirstName($ldapPbnlAccount->getCn());
+        $user->setSecondName($ldapPbnlAccount->getSn());
+        $user->setUidNumber(intval($ldapPbnlAccount->getUidNumber()));
+        $user->setMail($ldapPbnlAccount->getMail());
+        $user->setUsername($ldapPbnlAccount->getGivenName());
+        $user->setPostalCode($ldapPbnlAccount->getPostalCode());
+        $user->setMobilePhoneNumber($ldapPbnlAccount->getMobile());
+        $user->setStreet($ldapPbnlAccount->getStreet());
+        $user->generatePasswordAndSalt($ldapPbnlAccount->getUserPassword());
+        $user->setHomePhoneNumber($ldapPbnlAccount->getTelephoneNumber());
+
+        return $user;
```

Comment:
Gibt's in PHP soetwas, wie eine [tap-Methode](http://www.engineyard.com/blog/five-ruby-methods-you-should-be-using)?

Reply:
Ich verstehe noch nicht ganz, was diese hier bringen soll?

Reply:
Ich verstehe noch nicht ganz wofür man die hier braucht?